// @react-native-community/cli-platform-apple@14.1.0/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createBuild", {
  enumerable: true,
  get: function () {
    return _createBuild.default;
  }
});
Object.defineProperty(exports, "createLog", {
  enumerable: true,
  get: function () {
    return _createLog.default;
  }
});
Object.defineProperty(exports, "createRun", {
  enumerable: true,
  get: function () {
    return _createRun.default;
  }
});
Object.defineProperty(exports, "findPbxprojFile", {
  enumerable: true,
  get: function () {
    return _findPbxprojFile.default;
  }
});
Object.defineProperty(exports, "findPodfilePaths", {
  enumerable: true,
  get: function () {
    return _config.findPodfilePaths;
  }
});
Object.defineProperty(exports, "findXcodeProject", {
  enumerable: true,
  get: function () {
    return _findXcodeProject.default;
  }
});
Object.defineProperty(exports, "getArchitecture", {
  enumerable: true,
  get: function () {
    return _getArchitecture.default;
  }
});
Object.defineProperty(exports, "getBuildOptions", {
  enumerable: true,
  get: function () {
    return _buildOptions.getBuildOptions;
  }
});
Object.defineProperty(exports, "getDependencyConfig", {
  enumerable: true,
  get: function () {
    return _config.getDependencyConfig;
  }
});
Object.defineProperty(exports, "getLogOptions", {
  enumerable: true,
  get: function () {
    return _logOptions.getLogOptions;
  }
});
Object.defineProperty(exports, "getProjectConfig", {
  enumerable: true,
  get: function () {
    return _config.getProjectConfig;
  }
});
Object.defineProperty(exports, "getRunOptions", {
  enumerable: true,
  get: function () {
    return _runOptions.getRunOptions;
  }
});
Object.defineProperty(exports, "installPods", {
  enumerable: true,
  get: function () {
    return _installPods.default;
  }
});
var _config = require("./config");
var _buildOptions = require("./commands/buildCommand/buildOptions");
var _logOptions = require("./commands/logCommand/logOptions");
var _runOptions = require("./commands/runCommand/runOptions");
var _createBuild = _interopRequireDefault(require("./commands/buildCommand/createBuild"));
var _createLog = _interopRequireDefault(require("./commands/logCommand/createLog"));
var _createRun = _interopRequireDefault(require("./commands/runCommand/createRun"));
var _getArchitecture = _interopRequireDefault(require("./tools/getArchitecture"));
var _installPods = _interopRequireDefault(require("./tools/installPods"));
var _findXcodeProject = _interopRequireDefault(require("./config/findXcodeProject"));
var _findPbxprojFile = _interopRequireDefault(require("./config/findPbxprojFile"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
